agent/checks/docker.go

```diff
@@ -151,11 +158,17 @@ func (c *DockerClient) CreateExec(containerID string, cmd []string) (string, err
 }
 
 func (c *DockerClient) StartExec(containerID, execID string) (*circbuf.Buffer, error) {
-	data := struct{ Detach, Tty bool }{Detach: false, Tty: true}
+	data := struct{ Detach, Tty bool }{Detach: false, Tty: false}
 	uri := fmt.Sprintf("/exec/%s/start", execID)
 	b, code, err := c.call("POST", uri, data)
 	switch {
-	case err != nil:
+	// todo(fs): https://github.com/hashicorp/consul/pull/3621
+	// todo(fs): for some reason the docker agent closes the connection during the
+	// todo(fs): io.Copy call in c.call which causes a "connection reset by peer" error
+	// todo(fs): even though both body and status code have been received. My current is
+	// todo(fs): that the docker agent closes this prematurely but I don't understand why.
+	// todo(fs): the code below ignores this error.
+	case err != nil && !strings.Contains(err.Error(), "connection reset by peer"):
```

Comment:
If the container is shut down this is still a valid error condition and should fail the healthcheck. I would watch for this error and do a retry (maybe limit to 3 retries) rather than ignoring it altogether.

Reply:
The error comes from the docker agent and not the container. You can shut down the container and the check will go critical. I've tried that.

Reply:
What about if the docker agent daemon is shutdown and restarted after several minutes. Does the check still go critical? I believe it would because then this code will see 'connection refused' errors, not 'reset by peer' but want to confirm

Reply:
Good point. I'll test that

Reply:
I've tried this:

```
# run docker container
docker run -it ubuntu

# run consul
sudo /vagrant/consul agent -dev -config-dir /vagrant/conf.d/

$ cat conf.json
{
	"bind_addr": "127.0.0.1",
	"advertise_addr":"127.0.0.1",
	"bootstrap": true,
	"disable_remote_exec":false,
	"enable_script_checks":true,
	"node_id":"ac5ec7af-5238-4c18-988a-0385d0a0f477",
	"data_dir":"/tmp/consultest",
	"ports":{
		"https":-1,
		"serf_lan":9301
	},
	"datacenter":"mydc",
	"ui":true,
	"server":true
}
$ cat svc.json
{
	"service":{
		"name":"foo",
		"Check": {
			"name":"bar",
			"docker_container_id":"8afaa5f28703",
			"script":"echo hey",
			"interval":"2s"
		}
	}
}

# check healthy, no flaps

docker pause <id> -> check goes critical
docker unpause <id> -> check goes OK
docker kill <id> -> check goes critical and stays there

# start new container, update svc.json and pkill -HUP -f consul to reload
# check goes OK again

# stop docker agent
sudo service docker stop

# check goes critical "conn refused" since it never 
# gets to `StartExec` but fails one step before
```

Reply:
```ruby
$ cat Vagrantfile
Vagrant.configure("2") do |config|
  config.vm.define "ubuntu" do |c|
    c.vm.box = "ubuntu/xenial64"
    c.vm.hostname = "docker"
    c.vm.box_check_update = false
    c.vm.provision "shell", inline: <<-SHELL
	  apt-get install -y apt-transport-https ca-certificates curl software-properties-common
	  curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add -

	  apt-key fingerprint 0EBFCD88
	  add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
	  apt-get update

	  apt-get install -y docker-ce=17.03.1~ce-0~ubuntu-xenial
	  docker run hello-world
	  systemctl enable docker

	  groupadd docker
	  usermod -aG docker $USER
	  docker run hello-world

	  apt-get install -y unzip
	  wget -q https://releases.hashicorp.com/consul/1.0.0/consul_1.0.0_linux_amd64.zip
	  unzip -f consul_1.0.0_linux_amd64.zip
	  rm *.zip
    SHELL
  end
end
```

Reply:
I've built the consul binary with 

```sh
$ make linux && cp pkg/linux_amd64/consul ~/vagrant/ubuntu-docker
```

Reply:
:+1: